src/layers/pool/eigenpool.jl

```diff
@@ -0,0 +1,49 @@
+using Flux
+using Flux: glorot_uniform, normalise, @functor#, destructure
+using Zygote: @adjoint, @nograd
+using LinearAlgebra
+using ChemistryFeaturization
+
+#= Pooling Layer based on
+Graph Convolutional Networks with EigenPooling - Yao Ma, Suhang Wang, Charu C. Aggarwal, Jiliang Tang
+https://arxiv.org/abs/1904.13107
+
+Since the size of the matrices we are dealing with isn't as huge as the one in the original work,
+let us try using this as a global pooling mechanism instead, i.e., the adjacency matrix of the crystal
+itself would be only "subgraph" from which we "coarsen" to a single node and pool accordingly, which
+in theory would give us the overall graph representation
+=#
+
+# TBD - what other fields would be necessary for the pooling layer itself?
+struct EigenPool
+    pool::Function # pooling operator applied over the adjacency matrix
+    function EigenPool()
+        new(eigen_pooling)
+    end
+end
+
+# here, L is the laplacian matrix
+# this probably needs to be optimized.
+function eigen_pooling(L::Matrix{<:Real}, features::Matrix{<:Real})
+    L_eigen_vectors = eigvecs(L)    # find eigen vectors for L
+    result = Vector()
+
+    for i = 1:size(L_eigen_vectors)[1]
+        L_i = L_eigen_vectors[:, i]
+        push!(result, L_i'*features)
+    end
+
+    # using an agreeable H and then return H elements of result hcatt-ed into a single 1xdH vector
+    result = hcat(result...)'
+    reshape(result, length(result), 1)  # return it as a dHx1 Matrix
```

Comment:
Yeah, but returning something of length `N x D` isn't going to work because ultimately the size can't depend on `N` if you want to be able to feed in graphs of different sizes.

As per my comment above, I think a sensible way to do this could be that there's a parameter H that says how many eigenvectors to keep, then we can guarantee a return length of `H x D`.

OR, if we're worried about that having inconsistent performance across graph sizes, we could instead specify a parameter (say `h`) in (0,1) that says the _fraction_ of eigenvectors to keep and then do a standard (e.g. max or mean) pooling across a list of `h` vectors of length `D` to always return a vector of length `D`.

Reply:
> Yeah, but returning something of length N x D isn't going to work because ultimately the size can't depend on N if you want to be able to feed in graphs of different sizes.

Fair enough. We need to standardize across different graph sizes.

> As per my comment above, I think a sensible way to do this could be that there's a parameter H that says how many eigenvectors to keep, then we can guarantee a return length of H x D.

Would having `H::Integer` as a field in `EigenPool` be the best solution for this? I figured that `H` for a graph of size `N1` might not be appropriate (or rather, the "best") `H` for another graph of size `N2`.

Instead, what if, like `AGNPool`, we let users determine what they want their pooled feature length to be? Essentially,
`N * H` + `length(zero padding)` = `pooled_feature_length`

That way, pooled features would all be of similar lengths regardless of the graph sizes?

Reply:
Addressed in https://github.com/Chemellia/AtomicGraphNets.jl/pull/90/commits/a38fdf43a9aee58a8f2fd9d172efcbd9d329111d